open-format/subgraph
# Feature/of 99 feature index erc721 badge contract in subgraph

## Summary

Adds indexing functionality for erc721 badge contract
updates schema so badges have a `metadataURI` property when metadata for every token is the same.

## Description

The rational in change of schema:
We wanted to update the schema so `badges` have a `metadataURI` property and `badgeTokens` not to. However this will potentially cause breaking changes with all previous ERC721 contracts, and effectively won't index them correctly. So I opted for a compromise. 

The badgeToken metadataURI property can now be `null`, only set when each badge could have a separate metadata.
The badge metadataURI property is also nullable and only set on when the implementation id is "Badge"

This means when `setBaseURI` is called on a ERC721Badge contract it is only requires updating one string instead of every `badgeToken` on the `badge`

## Related Issue/Bounty

resolves [issue OF-99
](https://linear.app/openformat/issue/OF-99/feature-index-erc721-badge-contract-in-subgraph)

## Type of Change

Please check the boxes that apply to your pull request:


## Testing

I ran into the same issue as @nup9151f  mentioned when testing the subgraph and annoyingly couldn't get any meaningful unit tests to run

This is the long way round for testing

1. open up contracts repo
2. `anvil` or I found I had to run `anvil --host 0.0.0.0` 
3. in another terminal `make deploy`
4. `make CreateApp "hello"` and make sure APP_ID in .env matches the logged address
5. `make createERC721Badge rpc="anvil"` creates new badge contract
6. `make createERC721Base rpc="anvil"` creates new base contract
7. now spin up a local graph node cd to `docker` run `rm -rf data && docker-compose up --force-recreate` (I have to use sudo for this)
9. now in subgraph repo make sure the `/networks/local.json` matches the deploy AppFactory address (can find in the contracts repo `/deployed/31337.json`
10. `yarn prepare:local`
11. `yarn gen:local`
12. `yarn create:local`
13. `yarn deploy:local` and type version "v0.0.1"
14. if all that goes well you should be able to go to `http://0.0.0.0:8000/subgraphs/name/open-format/mumbai/graphql`
15. paste the query below and see the desired results. phewf 
```
query badges {
  badges (first:10){
    id
    name
    metadataURI
    tokens {
      id
      tokenId
      metadataURI
      owner {
        id
      }
      createdAt
      createdAtBlock
      updatedAt
      updatedAtBlock
    }
  }
}
```

You can also mint tokens from the contracts repo with:
`make ERC721Badge.mintTo args="<contract-address>"`
`make ERC721Base.mintTo args="<contract-address> somemetadatauri"` ( see https://github.com/open-format/contracts/pull/125)
make sure to update the contract addresses to the ones deployed

## Checklist

- [x] I have followed the project's [style guide](CONTRIBUTING.md#style-guide)
- [x] I have added or updated tests to cover my changes
- [x] I have checked that my changes do not introduce any new security vulnerabilities
- [x] I have provided a clear and concise description of my changes
- [x] I have linked to the related issue or bounty, if applicable

## Additional Information

`subgraph.local.yaml` and `local.json` get updated when running the scripts but I feel like these can be ignored
